car-bridge/src/drainage.rs

```diff
@@ -0,0 +1,72 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+
+use futures::Future;
+use tokio::{
+    spawn,
+    sync::mpsc::{self, Receiver, Sender},
+    task::JoinHandle,
+};
+
+/// Spawn tasks on the `Drainage` to ensure that the task completed when
+/// shutting down.
+pub struct Drainage {
+    drained_receiver: Receiver<()>,
+    drained_sender: Sender<()>,
+}
+
+impl Drainage {
+    pub fn new() -> Self {
+        let (drained_sender, drained_receiver) = mpsc::channel(1);
+        Self { drained_receiver, drained_sender }
+    }
+
+    /// Spawns a task while allowing to wait for it to complete when calling
+    /// `drain`.
+    pub fn spawn<F>(&self, f: F) -> JoinHandle<F::Output>
+    where
+        F: Future + Send + 'static,
+        F::Output: Send,
+    {
+        let sender = self.drained_sender.clone();
+
+        spawn(async move {
+            let result = f.await;
+            drop(sender);
+            result
+        })
+    }
+
+    /// Waits for all tasks that were spawned via this drainage to finish.
+    pub async fn drain(mut self) {
+        drop(self.drained_sender);
+        // As soon as every sender is dropped, the channel is closed and the
+        // receiver receives a message. Based on:
```

Comment:
I think you mean to say that the reception fails rather than getting a message?

```suggestion
        // reception fails. Based on:
```